core/ir/riscv64/codec.c

```diff
@@ -0,0 +1,1066 @@
+        case 1:
+            if ((bit11to7 == 0) && (bit6to2 == 0))
+                return &instr_infos[OP_c_ebreak];
+            else if (bit6to2 == 0)
+                return &instr_infos[OP_c_jalr];
+            else
+                return &instr_infos[OP_c_add];
+        default: return NULL;
+        }
+    case 5: return &instr_infos[OP_c_fsdsp];
+    case 6: return &instr_infos[OP_c_swsp];
+    case 7:
+        if (rv32)
+            return &instr_infos[OP_c_fswsp];
+        else if (rv64)
+            return &instr_infos[OP_c_sdsp];
+        else
+            return NULL;
+    default: return NULL;
+    }
+}
+
+static rv_instr_info_t *
+get_rvc_instr_info(uint32_t inst, int xlen)
+{
+    /* 0 is an illegal instruction which is often used as a canary */
+    if (inst == 0)
+        return &instr_infos[OP_unimp];
+
+    int op = inst & 0b11;
+    int funct = (inst >> 13) & 0b111;
+    int bit11to7 = (inst >> 7) & 0b11111;
+    int funct2 = (inst >> 10) & 0b11;
+    int bit12 = (inst >> 12) & 0b1;
+    int funct3 = (inst >> 5) & 0b11;
+    int bit6to2 = (inst >> 2) & 0b11111;
+    bool rv32 = xlen == 32;
+    bool rv64 = xlen == 64;
+    rv_instr_info_t *info = NULL;
+    uint32_t mask, match;
+
+    switch (op) {
+    case 0: info = match_op_0(funct, rv32, rv64); break;
+    case 1: info = match_op_1(funct, funct2, funct3, bit11to7, bit12, rv32, rv64); break;
+    case 2: info = match_op_2(funct, bit11to7, bit6to2, bit12, rv32, rv64); break;
+    }
+
+    if (info == NULL)
+        return NULL;
+    mask = info->nfo.code & 0xFFFFFFFF;
+    match = (info->nfo.code >> 32) & 0xFFFFFFFF;
+
+    if ((inst & mask) != match)
+        return NULL;
+    return info;
+}
+
+static rv_instr_info_t *
+get_rv_instr_info(uint32_t inst, trie_node_t trie[])
+{
+    rv_instr_info_t *info;
+    uint32_t mask, match;
+    trie_node_t *node;
+    size_t index;
+
+    /* We know the first index into trie straight from the instruction */
+    index = (inst & 0x7f) + 1;
+    while (index < (int16_t)-1) {
+        node = &trie[index];
+        ASSERT(node != NULL);
+        if (node->mask == 0) {
+            if (node->index >= ARRAY_SIZE(instr_infos))
+                return NULL;
+            info = &instr_infos[node->index];
+            mask = info->nfo.code & 0xFFFFFFFF;
+            match = (info->nfo.code >> 32) & 0xFFFFFFFF;
+            if ((inst & mask) != match)
+                return NULL;
+            return info;
+        }
+        index = node->index + ((inst >> node->shift) & node->mask);
+    }
+    return NULL;
+}
+
+/*
+ * End of format decoding functions.
+ **********************************************************/
+
+instr_info_t *
+get_instruction_info(uint opc)
+{
+    if (opc >= ARRAY_SIZE(instr_infos))
+        return NULL;
+    return &instr_infos[opc].nfo;
+}
+
+byte *
+decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
+{
+    /* Decode instruction width from the opcode */
```

Comment:
style nit: Prefer end punctuation in comments.

Reply:
Done